### DISCUSSION TITLE
QueryBuilder Insert with relations ?

### DISCUSSION THREAD

**Discussion by ElBambu** [mikro-orm/mikro-orm#3304] [Q&A]
Hello i'm using mikro-orm for a forum i'm building, i'm a front end guy so i'm pretty new with this

but basically i have this entity

```
@Entity()
export class Forum {
  @Index({ name: "forum_id" })
  @PrimaryKey()
  id: string = v4();

  @Property()
  title: string;

  @ManyToOne({ nullable: true })
  parentForum?: Forum;

  @OneToMany({
    entity: () => Forum,
    mappedBy: "parentForum",
    orphanRemoval: true,
  })
  childForum = new Collection<Forum>(this);

  @OneToMany("Thread", "forum", { orphanRemoval: true })
  threads = new Collection<Thread>(this);

  constructor(title: string) {
    this.title = title;
  }
}
```
which is an entity that references itself and can gave threads attached to it

and when i try to do an insert like this

```
    const forum = new Forum(title);
    (forum.parentForum as any) = parentId; // i also don't know if this is how it's supposed to be done
    await DI.em.createQueryBuilder(Forum).insert(forum).execute();
```

and i'm getting this error 

`insert into "forum" ("childForum", "id", "parent_forum_id", "threads", "title") values ('[object Object]', '6d225534-66c3-450d-972f-464abc3bef4c', NULL, '[object Object]', 'Forum Central') returning "id" - the column "childForum" of the "forum" relation does not exist`

i know it's because threads it's not supposed to be there but i don't know how to remove it xD

while if i do with the persistAndFlush it works ( i assume it's bc it does his mapping etc ) 

i could do it with the persistAndFlush but i'd like to stay a little closer to the sql ( for learning purposes )

**Comment by B4nan** [comment]
You should not put the entity instance into `qb.insert()` as that method expects POJO instead. That's probably the problem here, causing your `childForum` collection to be taken into account, while its a virtual property (a collection).

Try using `em.nativeInsert()` instead, as that expects an entity instance and will handle this properly, it will even try to handle collection updates for you.

```ts
await DI.em.nativeInsert(Forum, forum);
```

Or even better, just use the UoW as in next example:

> (forum.parentForum as any) = parentId; // i also don't know if this is how it's supposed to be done

This is indeed wrong, you need to put there an entity instance, not its PK. 

```ts
const forum = new Forum(title);
// get the entity instance without querying the database (alternatively you could do `await em.findOne()` which will query the db)
forum.parentForum = em.getReference(Forum, parentId); 
await em.persist(forum).flush();
```
